File: AlgorithmGoVersion/src/DataStructure/sort/innerSort/innerSortImpl/MergeSort.go
package innerSortImpl

/**
 * @author 蔚蔚樱
 * @version 1.0
 * @date 2021-06-22 00:09
 * @author—Email 蔚蔚樱first@outlook.com
 * @blogURL https://blog.csdn.net/ljfirst
 * @description 归并排序
 */

type MergeSort struct {
}

func (s *MergeSort) SortMethod(array []int) []int {

	if array == nil || len(array) <= 1 {
		return array
	}
	temp := make([]int, len(array))
	s.mergeSort(array, temp, 0, len(array)-1)
	return array
}

func (s *MergeSort) mergeSort(array, temp []int, left, right int) {
	if left < right {
		mid := (left + right) / 2
		s.mergeSort(array, temp, left, mid)
		s.mergeSort(array, temp, mid+1, right)
		s.sort(array, temp, left, mid, right)
	}
}

func (s *MergeSort) sort(array, temp []int, left, mid, right int) {
	l := left
	r := mid + 1
	position := left
	for l <= mid && r <= right {
		if array[l] < array[r] {
			temp[position] = array[l]
			l++
		} else {
			temp[position] = array[r]
			r++
		}
		position++
	}
	for l <= mid {
		temp[position] = array[l]
		l++
		position++
	}
	for r <= right {
		temp[position] = array[r]
		r++
		position++
	}
	//原址
	for i := left; i <= right; i++ {
		array[i] = temp[i]
	}
}

func (s *MergeSort) InterfaceName() string {
	return "ArraySortINTF"
}

File: AlgorithmJavaVersion/src/DataStructure/sort/innerSort/innerSortImpl/MergeSort.java
package DataStructure.sort.innerSort.innerSortImpl;

import Common.Utils.UTFactory;
import org.junit.Test;
import DataStructure.sort.innerSort.ArraySortINTF;

/**
 * @author 蔚蔚樱
 * @version 1.0
 * @date 2019-3-23 12:09:20
 * @author-Email micromicrohard@outlook.com
 * @description 归并排序
 * @URL
 */
public class MergeSort implements ArraySortINTF {

    @Test // 验证功能：用于全量测试
    public void TestFunc() throws Exception {
        UTFactory.FullTest(this.getClass());
    }

    
    public void sortMethod(int[] array) {
        if (!check(array)) {
            return;
        }
        int[] temp = new int[array.length];
        mergesort(array, 0, array.length - 1, temp);
    }

    public void mergesort(int[] array, int left, int right, int[] temp) {
        if (left < right) {
            int mid = (left + right) / 2;
            mergesort(array, left, mid, temp);
            mergesort(array, mid + 1, right, temp);
            sort(array, left, mid, right, temp);
        }
    }

    public void sort(int[] array, int left, int mid, int right, int[] temp) {

        int l = left;
        int r = mid + 1;
        int j = left;
        while (l <= mid && r <= right) {
            if (array[l] > array[r]) {
                temp[j++] = array[r++];
            } else {
                temp[j++] = array[l++];
            }
        }
        while (l <= mid) {
            temp[j++] = array[l++];
        }
        while (r <= right) {
            temp[j++] = array[r++];
        }
        if (right + 1 - left >= 0)
            System.arraycopy(temp, left, array, left, right + 1 - left);
    }
}
